libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Monad.hs

```diff
@@ -0,0 +1,181 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Test.Cardano.Ledger.Constrained.Monad (
+  Typed (..),
+  LiftT (..),
+  failT,
+  sameRep,
+  explain,
+  mergeExplain,
+  Dyn (..),
+  runDyn,
+  isDynSet,
+  HasCond (..),
+  hasOrd,
+  Id (..),
+  testHasCond,
+  errorTyped,
+  monadTyped,
+) where
+
+import Data.Set (Set)
+import Test.Cardano.Ledger.Constrained.Classes (SumCond (..))
+import Test.Cardano.Ledger.Constrained.TypeRep (
+  Proof (..),
+  Rep (..),
+  Size (..),
+  synopsis,
+  testEql,
+  (:~:) (Refl),
+ )
+
+-- =================================================
+
+newtype Typed x = Typed {runTyped :: Either [String] x}
+  deriving (Functor, Applicative, Monad)
+
+failT :: [String] -> Typed a
+failT ss = Typed (Left ss)
+
+class LiftT x where
+  liftT :: x -> Typed x
+  dropT :: Typed x -> x
+
+instance LiftT SumCond where
+  liftT (CondNever xs) = failT xs
+  liftT x = pure x
+  dropT (Typed (Left s)) = CondNever s
+  dropT (Typed (Right x)) = x
+
+instance LiftT Size where
+  liftT (SzNever xs) = failT xs
+  liftT x = pure x
+  dropT (Typed (Left s)) = SzNever s
+  dropT (Typed (Right x)) = x
+
+-- ==================================
+
+sameRep :: Rep era i -> Rep era j -> Typed (i :~: j)
+sameRep r1 r2 = case testEql r1 r2 of
+  Just x -> pure x
+  Nothing -> failT ["Type error in sameRep:\n  " ++ show r1 ++ " =/=\n  " ++ show r2]
+
+explain :: String -> Typed a -> Typed a
+explain s (Typed (Left ss)) = Typed (Left (s : ss))
+explain _ (Typed (Right x)) = Typed (Right x)
+
+mergeExplain :: (Monoid x, LiftT x) => String -> x -> x -> x
+mergeExplain message x y = dropT (explain message (liftT (x <> y)))
+
+-- ======================================================
+-- converting from a (Typed t) to something else
+
+-- The projection from (newtype Typed t = Typed {runTyped :: Either [String] x})
+-- runTyped :: Typed t => Either [String] x
+
+-- | Pushes the (Left msgs) into a call to 'error'
+errorTyped :: Typed t -> t
+errorTyped t = case runTyped t of
+  Right x -> x
+  Left xs -> error (unlines ("\nSolver-time error" : xs))
+
+-- | Pushes the (Left msgs) into a call to 'error', then injects into a Monad
+monadTyped :: Monad m => Typed t -> m t
+monadTyped t = pure (errorTyped t)
```

Comment:
It might be better to make it strict, otherwise it is possible to carry a thunk with bottom around, unless that is desired, of course.
```suggestion
monadTyped t = pure $! errorTyped t
```

Reply:
done